core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -180,10 +188,9 @@ private void publishDownstream(@Nullable U item, @Nullable Throwable cause) {
                         if (requestedByDownstream != Long.MAX_VALUE) {
                             requestedByDownstream--;
                         }
-                        state = State.REQUESTED;
+
+                        requestedFromUpstream++;
                         upstream.request(1);
```

Comment:
Should we request more than 1 if `maxConcurrency` is available?

Reply:
If my thinking is correct we should only have 1 available slot here, since this code will be executed each time a future finishes. Assuming downstream has requested some large amount we will always have `maxConcurrency == requestedFromUpstream + pendingFutures`. Since `pendingFutures` decreases by 1 here we only have space to increase `requestedFromUpstream` by 1.